File: android/app/src/test/java/fr/vferries/cuisine/data/ChipsTest.kt
package fr.vferries.cuisine.data

import org.junit.Assert.assertEquals
import org.junit.Test

class ChipsTest {

    private fun meta(
        slug: String,
        cuisine: String = "vietnamienne",
        course: String = "plat",
        difficulty: String = "moyenne",
        tags: List<String> = emptyList(),
        totalTime: Int = 30,
    ) = RecipeMeta(
        slug = slug,
        title = "t",
        servings = 2,
        prepTime = 10,
        cookTime = 20,
        totalTime = totalTime,
        difficulty = difficulty,
        cuisine = cuisine,
        course = course,
        tags = tags,
        updatedAt = "2026-04-20",
    )

    private val porc = meta("porc", totalTime = 45, tags = listOf("asiatique"))
    private val salade = meta("salade", cuisine = "italienne", course = "entrée", difficulty = "facile", tags = listOf("végé"), totalTime = 10)
    private val tarte = meta("tarte", cuisine = "française", course = "dessert", difficulty = "difficile", tags = listOf("dessert", "sans gluten"), totalTime = 60)

    @Test fun all_returns_all_slugs() {
        assertEquals(
            listOf("porc", "salade", "tarte"),
            filterByChip(listOf(porc, salade, tarte), ChipKey.ALL),
        )
    }

    @Test fun rapide_filters_on_totalTime_le_30() {
        assertEquals(listOf("salade"), filterByChip(listOf(porc, salade, tarte), ChipKey.RAPIDE))
    }

    @Test fun vege_filters_on_tag() {
        assertEquals(listOf("salade"), filterByChip(listOf(porc, salade, tarte), ChipKey.VEGE))
    }

    @Test fun asiatique_filters_on_tag() {
        assertEquals(listOf("porc"), filterByChip(listOf(porc, salade, tarte), ChipKey.ASIATIQUE))
    }

    @Test fun francais_filters_on_cuisine() {
        assertEquals(listOf("tarte"), filterByChip(listOf(porc, salade, tarte), ChipKey.FRANCAIS))
    }

    @Test fun dessert_filters_on_tag() {
        assertEquals(listOf("tarte"), filterByChip(listOf(porc, salade, tarte), ChipKey.DESSERT))
    }

    @Test fun favoris_filters_on_provided_set() {
        assertEquals(
            listOf("porc", "tarte"),
            filterByChip(
                listOf(porc, salade, tarte),
                ChipKey.FAVORIS,
                favorites = setOf("porc", "tarte"),
            ),
        )
    }

    @Test fun favoris_returns_empty_when_no_favorites() {
        assertEquals(
            emptyList<String>(),
            filterByChip(
                listOf(porc, salade, tarte),
                ChipKey.FAVORIS,
                favorites = emptySet(),
            ),
        )
    }

    @Test fun course_null_returns_all() {
        assertEquals(
            listOf("porc", "salade", "tarte"),
            filterByCourse(listOf(porc, salade, tarte), null),
        )
    }

    @Test fun course_entree_keeps_only_entrees() {
        assertEquals(
            listOf("salade"),
            filterByCourse(listOf(porc, salade, tarte), Course.ENTREE),
        )
    }

    @Test fun course_dessert_keeps_only_desserts() {
        assertEquals(
            listOf("tarte"),
            filterByCourse(listOf(porc, salade, tarte), Course.DESSERT),
        )
    }

    @Test fun difficulty_facile_keeps_only_faciles() {
        assertEquals(
            listOf("salade"),
            filterByDifficulty(listOf(porc, salade, tarte), Difficulty.FACILE),
        )
    }

    @Test fun sans_gluten_inactive_returns_all() {
        assertEquals(
            listOf("porc", "salade", "tarte"),
            filterBySansGluten(listOf(porc, salade, tarte), false),
        )
    }

    @Test fun sans_gluten_active_keeps_only_tagged() {
        assertEquals(
            listOf("tarte"),
            filterBySansGluten(listOf(porc, salade, tarte), true),
        )
    }
}

File: web/src/lib/chips.test.ts
import { describe, expect, it } from "vitest";
import {
  filterByChip,
  filterByCourse,
  filterByDifficulty,
  filterBySansGluten,
} from "./chips";

const porc = {
  slug: "porc-bigorre-caramel",
  cuisine: "vietnamienne",
  course: "plat",
  difficulty: "moyenne",
  tags: ["porc", "asiatique"],
  totalTime: 45,
};

const salade = {
  slug: "salade-tomate-mozza",
  cuisine: "italienne",
  course: "entrée",
  difficulty: "facile",
  tags: ["végé", "rapide"],
  totalTime: 10,
};

const tarte = {
  slug: "tarte-tatin",
  cuisine: "française",
  course: "dessert",
  difficulty: "difficile",
  tags: ["dessert"],
  totalTime: 60,
};

describe("filterByChip", () => {
  it("chip 'all' retourne tous les slugs", () => {
    expect(filterByChip([porc, salade], "all")).toEqual([
      "porc-bigorre-caramel",
      "salade-tomate-mozza",
    ]);
  });

  it("chip 'vege' ne retient que les recettes avec le tag 'végé'", () => {
    expect(filterByChip([porc, salade], "vege")).toEqual([
      "salade-tomate-mozza",
    ]);
  });

  it("chip 'rapide' ne retient que les recettes dont le total ≤ 30 min", () => {
    expect(filterByChip([porc, salade, tarte], "rapide")).toEqual([
      "salade-tomate-mozza",
    ]);
  });

  it("chip 'asiatique' ne retient que les recettes avec le tag 'asiatique'", () => {
    expect(filterByChip([porc, salade, tarte], "asiatique")).toEqual([
      "porc-bigorre-caramel",
    ]);
  });

  it("chip 'francais' ne retient que les recettes de cuisine française", () => {
    expect(filterByChip([porc, salade, tarte], "francais")).toEqual([
      "tarte-tatin",
    ]);
  });

  it("chip 'dessert' ne retient que les recettes avec le tag 'dessert'", () => {
    expect(filterByChip([porc, salade, tarte], "dessert")).toEqual([
      "tarte-tatin",
    ]);
  });

  it("chip 'favoris' ne retient que les recettes dont le slug est marqué favori", () => {
    expect(
      filterByChip([porc, salade, tarte], "favoris", { favorites: ["tarte-tatin", "porc-bigorre-caramel"] }),
    ).toEqual(["porc-bigorre-caramel", "tarte-tatin"]);
  });

  it("chip 'favoris' retourne [] si aucun favori connu", () => {
    expect(filterByChip([porc, salade, tarte], "favoris", { favorites: [] })).toEqual([]);
  });
});

describe("filterByCourse", () => {
  it("retourne tous les slugs quand aucune course n'est sélectionnée", () => {
    expect(filterByCourse([porc, salade, tarte], null)).toEqual([
      "porc-bigorre-caramel",
      "salade-tomate-mozza",
      "tarte-tatin",
    ]);
  });

  it("ne retient que les recettes dont course = 'entrée'", () => {
    expect(filterByCourse([porc, salade, tarte], "entrée")).toEqual([
      "salade-tomate-mozza",
    ]);
  });

  it("ne retient que les recettes dont course = 'plat'", () => {
    expect(filterByCourse([porc, salade, tarte], "plat")).toEqual([
      "porc-bigorre-caramel",
    ]);
  });

  it("ne retient que les recettes dont course = 'dessert'", () => {
    expect(filterByCourse([porc, salade, tarte], "dessert")).toEqual([
      "tarte-tatin",
    ]);
  });
});

describe("filterByDifficulty", () => {
  it("retourne tous les slugs quand aucune difficulté n'est sélectionnée", () => {
    expect(filterByDifficulty([porc, salade, tarte], null)).toEqual([
      "porc-bigorre-caramel",
      "salade-tomate-mozza",
      "tarte-tatin",
    ]);
  });

  it("ne retient que les recettes dont difficulty = 'facile'", () => {
    expect(filterByDifficulty([porc, salade, tarte], "facile")).toEqual([
      "salade-tomate-mozza",
    ]);
  });
});

describe("filterBySansGluten", () => {
  const flan = {
    slug: "flan",
    cuisine: "française",
    course: "dessert",
    difficulty: "facile",
    tags: ["dessert", "sans gluten"],
    totalTime: 80,
  };

  it("inactif: retourne tous les slugs", () => {
    expect(filterBySansGluten([porc, flan], false)).toEqual([
      "porc-bigorre-caramel",
      "flan",
    ]);
  });

  it("actif: ne retient que les recettes taguées 'sans gluten'", () => {
    expect(filterBySansGluten([porc, flan], true)).toEqual(["flan"]);
  });
});

describe("filterByChip — saison", () => {
  const doc = (slug: string, saisonMonths: number[]): ChipFilterDoc => ({
    slug,
    cuisine: "",
    tags: [],
    totalTime: 0,
    saisonMonths,
  });

  it("garde les recettes dont la plage contient le mois injecté", () => {
    const recipes = [doc("ete", [6, 7, 8]), doc("hiver", [12, 1, 2])];
    expect(filterByChip(recipes, "saison", { month: 7 })).toEqual(["ete"]);
    expect(filterByChip(recipes, "saison", { month: 1 })).toEqual(["hiver"]);
  });

  it("exclut une recette sans saisonMonths quand un mois est fourni", () => {
    const bare: ChipFilterDoc = { slug: "x", cuisine: "", tags: [], totalTime: 0 };
    expect(filterByChip([bare], "saison", { month: 7 })).toEqual([]);
  });

  it("sans mois fourni, ne filtre rien", () => {
    const recipes = [doc("ete", [6, 7, 8])];
    expect(filterByChip(recipes, "saison", {})).toEqual(["ete"]);
  });
});
